WordPress/openverse
# Add distributed reindex steps

## Fixes



Fixes #4148 by @stacimc 

## Description



Adds steps to the new data refresh DAGs to create the temporary ES index (which will later be promoted), and populating it via the distributed reindex:

<img width="664" alt="Screenshot 2024-08-19 at 4 30 34 PM" src="https://github.com/user-attachments/assets/26184991-71ba-4b12-b9ae-d90cd8c26351">

## Testing Instructions


`./ov just api/init && ./ov just up` and then run a few DAGs in order to ingest some new data (I used Jamendo for audio and Cleveland for image, and manually marked them as success as soon as some records were recorded). Then run the new data refresh DAGs (eg `staging_audio_data_refresh`). They should run successfully, but skip the tasks to create/terminate EC2 instances. Go to [elasticvue](https://app.elasticvue.com/cluster/0/indices) and you should see a new index matching the output of the `generate_index_name` task.

It should have no aliases. If you refresh the index in Elasticvue, you'll also see that it has the expected number of records (so for example, if you started from sample data the main audio index will have 5000 records. If you ingested 100 records from Jamendo before running the data refresh, your new temp index will have 5100 records.)

Run `./ov just api/pgcli` and `select count(*) from audio` (or image, depending on which data refresh you ran). Then verify the count is the same as the number of docs in the new index in Elasticvue.

Also note that you'll see somewhat unexpected behavior if you run the DAGs multiple times in a row right now, because we don't yet delete the temp table. That's expected! If you need to run it a second time, just drop the temp import tables manually in the API DB first.

It's not simple to test this against the staging environment when running locally because of the requirement that we skip these tasks locally, but I did so by configuring the `AIRFLOW_CONN_AWS_DEFAULT` locally and commenting out the logic to skip. This should technically be safe even in production because no data is destroyed or promoted, but if you do want to test this you should ONLY run the staging DAGs to be safe, and we should clean up the dangling tables/indices afterward. There are additional issues in the milestone for testing this formally in those environments so I do not think it is a requirement for testers here.

## Checklist


      PRs) or the media properties generator (`./ov just catalog/generate-docs media-props`
      for the catalog or `./ov just api/generate-docs` for the API) where applicable.

[best_practices]:
  https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin


<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>
